File: _site/python_solutions/659.split-array-into-consecutive-subsequences.py
#
# @lc app=leetcode id=659 lang=python
#
# [659] Split Array into Consecutive Subsequences
#
# https://leetcode.com/problems/split-array-into-consecutive-subsequences/description/
#
# algorithms
# Medium (38.53%)
# Total Accepted:    15.2K
# Total Submissions: 39.4K
# Testcase Example:  '[1,2,3,3,4,5]'
#
# You are given an integer array sorted in ascending order (may contain
# duplicates), you need to split them into several subsequences, where each
# subsequences consist of at least 3 consecutive integers. Return whether you
# can make such a split.
# 
# Example 1:
# 
# Input: [1,2,3,3,4,5]
# Output: True
# Explanation:
# You can split them into two consecutive subsequences : 
# 1, 2, 3
# 3, 4, 5
# 
# 
# 
# Example 2:
# 
# Input: [1,2,3,3,4,4,5,5]
# Output: True
# Explanation:
# You can split them into two consecutive subsequences : 
# 1, 2, 3, 4, 5
# 3, 4, 5
# 
# 
# 
# Example 3:
# 
# Input: [1,2,3,4,4,5]
# Output: False
# 
# 
# 
# Note:
# 
# The length of the input is in range of [1, 10000]
# 
# 
#
        
class Solution:
    def isPossible(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        freqs = Counter(nums)
        terminates = Counter()
        for n in nums:
            if freqs[n] == 0:
                continue
            freqs[n] -= 1

            if n - 1 in terminates and terminates[n - 1] > 0:
                terminates[n - 1] -= 1
                terminates[n] += 1
            elif freqs[n + 1] > 0 and freqs[n + 2] > 0:
                freqs[n + 1] -= 1
                freqs[n + 2] -= 1
                terminates[n + 2] += 1
            else:
                return False
        return True


nums = [1, 2, 3, 3, 4, 4, 5, 5]
print(Solution().isPossible(nums))

File: _site/python_solutions/659.split-array-into-consecutive-subsequences.rb
# @param {Integer[]} nums
# @return {Boolean}
def is_possible(nums)
  freqs = nums.each_with_object(Hash.new(0)) { |w, h| h[w] += 1; }
  terminates = Hash.new(0)
  nums.each do |n|
    next if freqs[n].zero?

    freqs[n] -= 1
    if terminates[n - 1] > 0
      terminates[n - 1] -= 1
      terminates[n] += 1
    elsif freqs[n + 1] > 0 && freqs[n + 2] > 0
      freqs[n + 1] -= 1
      freqs[n + 2] -= 1
      terminates[n + 2] += 1
    else
      return false
    end
  end
  true
end

nums = [1, 2, 3, 3, 4, 4, 5, 5]
nums = [1, 2, 3, 3, 4, 5]
# require 'random'
# nums = 100000.times.map {Random.rand(10)}
# p nums
p is_possible(nums)
